### ISSUE TITLE
Section 8.1.5 connection group transmission scheduler

### ISSUE THREAD

**Issue by mwelzl** [ietf-tapswg/api-drafts#1194]
_From Genart review: https://mailarchive.ietf.org/arch/msg/taps/L7_Aopnvp-xn9aEQ_zI9kgxts68/_

Is this SCTP specific or could it apply more generally?

**Comment by mwelzl** [comment]
Searching for the word "scheduler" in the MP-DCCP draft gives me only: "Therefore, the path priority should be considered as hints for the packet scheduler when making decisions which path to use for user plane traffic."

Yes we can have priorities - we have them!  but this is about a more fine-grain choice, to select the type of scheduler to use.

See, RFC 8260 offers a choice between:

First-Come, First-Served Scheduler (SCTP_SS_FCFS)
Round-Robin Scheduler (SCTP_SS_RR)
Round-Robin Scheduler per Packet (SCTP_SS_RR_PKT)
Priority-Based Scheduler (SCTP_SS_PRIO)
Fair Capacity Scheduler (SCTP_SS_FC)
Weighted Fair Queueing Scheduler (SCTP_SS_WFQ)

Should we just put this list there, as a choice, without the SCTP_* abbreviations?   I mean, these are pretty general schedulers. I'm not sure any protocol other than SCTP will ever allow people to choose the scheduling algorithm, really... so an alternative could be to make this SCTP-specific?   But I guess if we offer it as a general thing, with a list of schedulers, and all is just "best effort" anyway, nothing terrible will happen if a choice cannot be honored...

**Comment by gorryfair** [comment]
The text says:
"This property specifies which scheduler should be used among
   Connections within a Connection Group, see Section 7.4.  The set of
   schedulers can be taken from [RFC8260]."

I'm not saying cite MP-DCCP, or MP-QUIC,  but I agree these are generic methods, albeit explicitly stated in RFC8260. The list is different in  RFC 7567 Sect 2.1, but then  the discussion focus is slightly different, so I am not sure that a definitive list is necessarily the correct thing, but these are mostly generic. I note that WFQ, HTB (and others) need params... etc. 

Is this example of text any help at heading somewhere:

"This property specifies a preference for a scheduler among the
   Connections within a Connection Group, see Section 7.4.  This set of
   schedulers include:
First-Come, First-Served Scheduler (SS_FCFS)
Round-Robin Scheduler (SS_RR)
Priority-Based Scheduler (SS_PRIO)
Fair Capacity Scheduler (SS_FC)
(see examples in [RFC8260])."

**Comment by mirjak** [comment]
I not sure that correct. I think the list of schedulers can be different for each transport. For SCTP we have an RFC, so the list should be right but for other transports we don't have any such references. So we can give this list as examples but I don't think it nessecarily the same list for all transports.

**Comment by mwelzl** [comment]
We certainly don't have all these schedulers available in other protocols, no. But this is "best effort", just like the connection or message priorities and many other things. If it doesn't work, the application just won't really get the performance (including prioritization) that was desired.

We need to offer a concrete list though, just like we provide "unordered delivery allowed" while some protocols don't support this.

**Comment by abrunstrom** [comment]
Just to note that this is about connection (stream) scheduling and not path scheduling. I saw MP-DCCP and MP-QUIC mentioned above, but this is not about multipath. So this is mainly applicable only to SCTP and QUIC. We do not have a TAPS mapping for QUIC yet, so in that sense it seems ok to base it on SCTP.

**Comment by gorryfair** [comment]
So, before I got this diverted to talk about schedulers in general (which may be didn't help); I suggested something that was starting to look like:

"   This property specifies which scheduler should be used among
   Connections within a Connection Group, see Section 7.4.  A set of
   schedulers is described in [RFC8260]."

- Which is only a very small variation in wording. There isn't a QUIC spec for this, and I am not sure we need to say more.
